tools/workspace-plugin/src/executors/build/lib/babel.ts

```diff
@@ -0,0 +1,125 @@
+ */
+
+import { writeFile, readFile } from 'node:fs/promises';
+import { basename, join } from 'node:path';
+
+import { type BabelFileResult, transformAsync } from '@babel/core';
+import { globSync } from 'fast-glob';
+import { logger } from '@nx/devkit';
+
+import { processAsyncQueue, type NormalizedOptions } from './shared';
+import { compileSwc } from './swc';
+
+const EOL_REGEX = /\r?\n/g;
+
+function addSourceMappingUrl(code: string, sourceMapLocation: string): string {
+  // Babel keeps stripping this comment, even when correct option is set. Adding manually.
+  return code + '\n//# sourceMappingURL=' + sourceMapLocation;
+}
+
+export function hasStylesFilesToProcess(normalizedOptions: NormalizedOptions) {
+  const files = globSync('**/*.styles.ts', { cwd: normalizedOptions.absoluteSourceRoot });
+  return files.length > 0;
+}
+
+export async function compileWithGriffelStylesAOT(options: NormalizedOptions, successCallback: () => Promise<boolean>) {
+  const { esmConfig, restOfConfigs } = options.moduleOutput.reduce<{
+    esmConfig: NormalizedOptions['moduleOutput'][number] | null;
+    restOfConfigs: NormalizedOptions['moduleOutput'];
+  }>(
+    (acc, outputConfig) => {
+      if (outputConfig.module === 'es6') {
+        acc.esmConfig = outputConfig;
+        return acc;
+      }
+
+      acc.restOfConfigs.push(outputConfig);
+      return acc;
+    },
+    { esmConfig: null, restOfConfigs: [] },
+  );
+
+  if (!esmConfig) {
+    logger.warn('es6 module output not specified. Skipping griffel AOT...');
+    const compilationQueue = restOfConfigs.map(outputConfig => {
+      return compileSwc(outputConfig, options);
+    });
+    return processAsyncQueue(compilationQueue, successCallback);
+  }
+
+  await compileSwc(esmConfig, options);
+  await babel(esmConfig, options);
+
+  const compilationQueue = restOfConfigs.map(outputConfig => {
+    const overriddenSourceRoot = join(options.workspaceRoot, options.project.root);
+    // we need to override source root to the output path of transpiled ESM+Griffel AOT, because griffel is unable to handle SWC commonjs output
+    // so instead of transpiling TS(ESM) -> JS(COMMONJS), we transpile JS(ESM + griffel AOT) -> JS(COMMONJS)
+    const overriddenAbsoluteSourceRoot = join(overriddenSourceRoot, esmConfig.outputPath);
+
+    return compileSwc(outputConfig, {
+      ...options,
+      absoluteSourceRoot: overriddenAbsoluteSourceRoot,
+    });
+  });
+
+  return processAsyncQueue(compilationQueue, successCallback);
+}
+
+async function babel(esmModuleOutput: NormalizedOptions['moduleOutput'][number], normalizedOptions: NormalizedOptions) {
+  const filesRoot = join(normalizedOptions.absoluteProjectRoot, esmModuleOutput.outputPath);
+  const files = globSync('**/*.styles.js', { cwd: filesRoot });
+
+  if (files.length === 0) {
+    return;
+  }
+
+  logger.log(`Processing griffel AOT with babel: ${files.length} files`);
+
+  for (const filename of files) {
```

Comment:
question: won't it be better to create array of promises and later Promise.all? 

smth like this: 

```ts
  const processFiles = files.map(async filename => {
    const filePath = join(filesRoot, filename);

    try {
      const sourceCode = (await readFile(filePath)).toString().replace(EOL_REGEX, '\n');
      const result = (await transformAsync(sourceCode, {
        ...BABEL_OPTIONS,
        // to avoid leaking of global configs
        babelrcRoots: [normalizedOptions.absoluteProjectRoot],
        filename: filePath,
        sourceFileName: basename(filename),
      })) /* Bad `transformAsync` types. it can be null only if 2nd param is null(config)*/ as NonNullableRecord<BabelFileResult>;

      if (result.code === sourceCode) {
        logger.verbose(`babel: skipped ${filePath}`);
      } else {
        const sourceMapFile = filePath + '.map';

        await writeFile(filePath, result.code);
        await writeFile(sourceMapFile, JSON.stringify(result.map));

        logger.verbose(`babel: transformed ${filePath}`);
      }
    } catch (error) {
      logger.error(`Error processing file: ${filePath}: ${error}`);
    }
  });

  await Promise.all(processFiles);
}
```

Reply:
yeah Promise.all makes more sense here, not gonna change anything here -> this is implementation was copied from https://github.com/microsoft/fluentui/blob/master/scripts/tasks/src/babel.ts with some needed tweaks for `transformAsync` parameters.

also as you can see at the beginning of this module, this is gonna be removed completely ( hopefully soon )